Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift

```diff
@@ -561,34 +566,140 @@ final class AppDelegateViewModelTests: TestCase {
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.notDetermined)
 
-      self.registerUserNotificationSettings.assertValueCount(3)
+      self.getNotificationAuthorizationStatus.assertValueCount(2)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(1)
+      self.unregisterForRemoteNotifications.assertValueCount(0)
+
+      self.vm.inputs.notificationAuthorizationCompleted()
+
+      self.getNotificationAuthorizationStatus.assertValueCount(3)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(1)
+      self.unregisterForRemoteNotifications.assertValueCount(0)
+
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.authorized)
+
+      self.getNotificationAuthorizationStatus.assertValueCount(3)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(2)
       self.unregisterForRemoteNotifications.assertValueCount(0)
     }
 
     self.vm.inputs.userSessionEnded()
 
-    self.registerUserNotificationSettings.assertValueCount(3)
+    self.registerForRemoteNotifications.assertValueCount(2)
     self.unregisterForRemoteNotifications.assertValueCount(1)
   }
 
+  fileprivate func count(_ string: String) -> ([String]) -> Int {
+    return { array in
+      return array.reduce(0, { acc, val in val == string ? acc + 1 : acc })
+    }
+  }
+
+  func testTrackingPushAuthorizationOptIn() {
+    let client = MockTrackingClient()
+
+    withEnvironment(currentUser: .template, koala: Koala(client: client)) {
+      self.vm.inputs.userSessionStarted()
+
+      XCTAssertEqual(0, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.applicationDidEnterBackground()
+      self.vm.inputs.applicationWillEnterForeground()
+
+      XCTAssertEqual(0, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.notDetermined)
+
+      XCTAssertEqual(1, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.notificationAuthorizationCompleted()
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.authorized)
+
+      XCTAssertEqual(1, client.events |> count("Triggered Push Permissions Dialog"))
```

Comment:
@bormind VMs account for a lot of complete, composed logic, so I think of them as autonomous units themselves. This means we care about all changes.

I do understand the concern around verboseness and noise vs. signal per test, but there are some other approaches we could consider to alleviate this pain without sacrificing the information lost by the counting approach. We could, for example, assert on just new events after a certain point:

``` swift
let endIndex = client.events.endIndex

self.vm.inputs.notificationAuthorizationCompleted()
self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.authorized)

XCTAssertEqual(["Confirmed Push Opt-In"],
               client.events.suffix(from: endIndex)
```

It's cleaner, but we run the risk of not testing certain events.

Meanwhile, [Jest](http://facebook.github.io/jest/) has the ability to snapshot data structures, which—were we to build something like it—could provide that additional layer of checking with less boilerplate. In the end, that time investment hasn't seemed worth it (yet) and we generally embrace boilerplate if it gives us more confidence in the correctness of our code.